src/test/java/io/aiven/connect/elasticsearch/ElasticsearchSinkFailureTest.java

```diff
@@ -0,0 +1,192 @@
+
+import static org.junit.Assert.fail;
+
+public class ElasticsearchSinkFailureTest {
+
+    private static class MockedJestClient implements JestClient {
+        private final BulkResult success;
+
+        public MockedJestClient(final BulkResult success) {
+            this.success = success;
+        }
+
+        @Override
+        public void close() throws IOException {
+        }
+
+        @Override
+        public <T extends JestResult> T execute(final Action<T> clientRequest) throws IOException {
+            return (T) success;
+        }
+
+        @Override
+        public <T extends JestResult> void executeAsync(final Action<T> clientRequest,
+                                                        final JestResultHandler<? super T> jestResultHandler) {
+            jestResultHandler.completed((T) success);
+        }
+
+        @Override
+        public void shutdownClient() {
+
+        }
+
+        @Override
+        public void setServers(final Set<String> servers) {
+
+        }
+    }
+
+    abstract class CustomJestElasticsearchClient extends JestElasticsearchClient {
+        public final AtomicInteger apiCallCounter = new AtomicInteger(0);
+
+        public CustomJestElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public Version getVersion() {
+            return Version.ES_V7;
+        }
+
+        @Override
+        public JsonObject getMapping(final String index, final String type) {
+            return new JsonObject();
+        }
+    }
+
+    class CustomJestTemporarilyFailingElasticsearchClient extends CustomJestElasticsearchClient {
+        public CustomJestTemporarilyFailingElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public BulkResponse executeBulk(final BulkRequest bulk) throws IOException {
+            final int numAttempt = apiCallCounter.incrementAndGet();
+
+            if (numAttempt < 3) {
+                return BulkResponse.failure(true, "");
+            }
+
+            return BulkResponse.success();
+        }
+    }
+
+    class CustomJestFailingElasticsearchClient extends CustomJestElasticsearchClient {
+        public CustomJestFailingElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public BulkResponse executeBulk(final BulkRequest bulk) throws IOException {
+            apiCallCounter.incrementAndGet();
+            return BulkResponse.failure(false, "");
+        }
+    }
+
+    @Test
+    public void testRetryIfRecoverable() {
+        final BulkResult success = new BulkResult(new Gson());
+        success.setSucceeded(true);
+
+
+        final ElasticsearchSinkTask elasticsearchSinkTask = new ElasticsearchSinkTask();
+        final CustomJestElasticsearchClient failingClient =
+            new CustomJestTemporarilyFailingElasticsearchClient(new MockedJestClient(success));
```

Comment:
nit: Could we make more explicit the number of retries on this test? Either by mocking and setting the behavior on the same test method, or by adding an argument to the custom class should help.